plugins/thunderbolt/fu-thunderbolt-image.c

```diff
@@ -0,0 +1,571 @@
+	g_autoptr(GByteArray) read = g_byte_array_new ();
+
+	if (location_start > fw->len || location_start + location->len > fw->len) {
+		g_set_error (error,
+			     FWUPD_ERROR, FWUPD_ERROR_READ,
+			     "Given location is outside of the given FW (%s)",
+			     location->description);
+		return NULL;
+	}
+
+	read = g_byte_array_append (read,
+				    fw->data + location_start,
+				    location->len);
+
+	if (location->mask)
+		read->data[0] &= location->mask;
+	return g_steal_pointer (&read);
+}
+
+static guint32
+read_farb_pointer_impl (const FuThunderboltFwLocation  *location,
+			const FuThunderboltFwObject    *fw,
+			GError                        **error)
+{
+	guint32 value = 0;
+	g_autoptr(GByteArray) farb = read_location (location, fw, error);
+	if (farb == NULL)
+		return 0;
+	memcpy(&value, farb->data, farb->len);
+	return GUINT32_FROM_LE (value);
+}
+
+static guint32
+read_farb_pointer (const FuThunderboltFwObject *fw, GError **error)
+{
+	const FuThunderboltFwLocation farb1 = { .offset = 0,      .len = 3 };
+	const FuThunderboltFwLocation farb2 = { .offset = 0x1000, .len = 3 };
+
+	guint32 ret = read_farb_pointer_impl (&farb1, fw, error);
+	if (*error != NULL)
+		return ret;
+	if (valid_farb_pointer (ret))
+		return ret;
+
+	ret = read_farb_pointer_impl (&farb2, fw, error);
+	if (!valid_farb_pointer (ret)) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "Invalid FW image file format");
+		return 0;
+	}
+
+	return ret;
+}
+
+static gboolean
+compare (const FuThunderboltFwLocation  *location,
+	 const FuThunderboltFwObject    *controller_fw,
+	 const FuThunderboltFwObject    *image_fw,
+	 gboolean                       *result,
+	 GError                        **error)
+{
+	g_autoptr(GByteArray) controller_data = NULL;
+	g_autoptr(GByteArray) image_data      = NULL;
+
+	controller_data = read_location (location, controller_fw, error);
+	if (controller_data == NULL)
+		return FALSE;
+
+	image_data = read_location (location, image_fw, error);
+	if (image_data == NULL)
+		return FALSE;
+
+	*result = memcmp (controller_data->data, image_data->data, location->len) == 0;
+	return TRUE;
+}
+
+static void
+read_bool (const FuThunderboltFwLocation  *location,
+	   const FuThunderboltFwObject    *fw,
+	   gboolean                       *val,
+	   GError                        **error)
+{
+	g_autoptr(GByteArray) read = read_location (location, fw, error);
+	if (read == NULL)
+		return;
+	for (gsize i = 0; i < read->len; i++)
+		if (read->data[i] != 0) {
+			*val = TRUE;
+			return;
+		}
+	*val = FALSE;
+	return;
+}
+
+static guint16
+read_uint16 (const FuThunderboltFwLocation  *location,
+	     const FuThunderboltFwObject    *fw,
+	     GError                        **error)
+{
+	guint16 value = 0;
+	g_autoptr(GByteArray) read = read_location (location, fw, error);
+	g_assert_cmpuint (location->len, ==, sizeof (guint16));
+	if (read == NULL)
+		return 0;
+
+	memcpy(&value, read->data, read->len);
```

Comment:
missing a space after `memcpy `and before `(`.
This shows up a few other places too.

Reply:
I was sure I cleared those up. :(
Will go over it again.

Reply:
Don't worry, I've participated in fwupd development for a while now and I mess them up constantly too.  If there is some style checker tool we can put in CI that catches stuff like this on PR's I would love to integrate that.

Reply:
OK, you made me to uncover my cards earlier...
I planned to create a clang-format config file that will match the project style and send a PR.
At first, it helps developers by automating the formatting. It usually becomes a matter of pressing a key-combination, after a simple configuration of your favorite code editor or IDE.
Later, it can be integrated to git (with commit hook) and of course can run in CI (and test if any change was done).

Reply:
(There are other code formatting tools, but seems like most of them aren't maintained, that clang-format is the only modern one among them and it is highly configurable so it should be feasible to match the current style as much as possible.)

Reply:
clang-format works for me. Sounds good!

Reply:
You could use [this](https://gist.github.com/gicmo/eedf610d6e67dcc40f3a1e5a6d31c5e2) as a starting point. I used it to convert some source to a fwupd-like project.